### PR TITLE
Implement linear_ramp pad mode in jax.numpy.pad

### PR DESCRIPTION
**Pull request by minoring**
Related #5010

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -2347,6 +2347,31 @@ def _pad_edge(array, pad_width):
...
+            stop=edge.squeeze(axis), # Dimension is replaced by linspace
+            num=width,
+            endpoint=False,
+            dtype=array.dtype,
+            axis=axis
+        )
+        for end_value, edge, width in zip(
+            end_values[axis], edges, pad_width[axis]
+        )
+    )
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:2366]: Slight preference here for just writing two linspace operations; reading over this, it took me far longer to understand what the list comprehension was doing than it would have if the two statements were just written out.

Comment (minoring) [reply jax/_src/numpy/lax_numpy.py:2369]: Thanks for your review. I agree it was hard to understand.
I converted it into two statements. PTAL :)

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by gnecula** [comment]
When trying to import this we ran into test failures, on CPU/GPU/TPU. E.g., on CPU `LaxBackedNumpyTests.testPadLinearRamp_shape=int32[1,4]_mode=linear_ramp_pad_width=(2, 0)_end_values=10.0`
fails with 
```
Max relative difference: nan
 x: array([[10, 10, 10, 10, 10, 10],
       [10,  6,  3,  6,  5,  4],
       [10,  3, -4,  1,  0, -1]], dtype=int32)
 y: array([[10, 10, 10, 10, 10, 10],
       [10,  6,  3,  5,  5,  4],
       [10,  3, -4,  1,  0, -1]], dtype=int32)
```

I am wondering why this test did not fail in GitHub. It may be because google uses internally a different version of numpy (1.16.4)?

**Comment by jakevdp** [comment]
I suspect what's happening here is that rounding is happening in a different way for integer values. For example, here are two equivalent ways of generating float64 ramps:
```python
>>> np.linspace(0, 2, num=5, dtype=float)[::-1]                                                              
array([2. , 1.5, 1. , 0.5, 0. ])

>>> 2 - np.linspace(0, 2, num=5, dtype=float)                                                                
array([2. , 1.5, 1. , 0.5, 0. ])
```
but for integer dtypes, these equivalent approaches lead to differences in rounding direction:
```python
>>> np.linspace(0, 2, num=5, dtype=int)[::-1]                                                                
array([2, 1, 1, 0, 0])

>>> 2 - np.linspace(0, 2, num=5, dtype=int)                                                                  
array([2, 2, 1, 1, 0])
```
I suspect the numpy versions are generating linspaces one way, and the jax version is generating them the other way.

**Comment by minoring** [comment]
It seems that there was a bug in linear_ramp padding for numpy <= 1.15
https://github.com/numpy/numpy/commit/1c45e0df150b1f49982aaa3fc1a328407b5eff7e#

Here are test results I found.

"numpy 1.15.x"
```python
a = [[-4, 1, 0, -1]]
np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0)
array([[10, 10, 10, 10, 10, 10],
       [10,  6,  3,  6,  5,  4],
       [10,  3, -4,  1,  0, -1]])

np.linspace(start=10.0, stop=1, endpoint=False, num=2, dtype=int)
array([10,  5])
```

"numpy 1.16.x"
```python
a = [[-4, 1, 0, -1]]
np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0)
array([[10, 10, 10, 10, 10, 10],
       [10,  6,  3,  5,  5,  4],
       [10,  3, -4,  1,  0, -1]])

np.linspace(start=10.0, stop=1, endpoint=False, num=2, dtype=int)
array([10,  5])
```

`np.linspace` works in the same way, but not `np.pad`

I am going to remove numpy <= 1.15 when testing linear_ramp.

**Comment by jakevdp** [comment]
I think the issue must be something different - the tests in question were run on numpy 1.16.4

**Comment by jakevdp** [comment]
It looks like this issue was fixed in numpy version 1.16.6:
```python
>>> import numpy as np
>>> print(np.__version__)
1.16.6
>>> a = np.array([[-4, 1, 0, -1]], dtype='int32')
>>> print(np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0))
print(np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0))
[[10 10 10 10 10 10]
 [10  6  3  5  5  4]
 [10  3 -4  1  0 -1]]
```
```python
>>> import numpy as np
>>> print(np.__version__)
1.16.5
>>> a = np.array([[-4, 1, 0, -1]], dtype='int32')
>>> print(np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0))
print(np.pad(a, pad_width=(2, 0), mode='linear_ramp', end_values=10.0))
[[10 10 10 10 10 10]
 [10  6  3  6  5  4]
 [10  3 -4  1  0 -1]]
```

**Comment by minoring** [comment]
Yes, that seems the answer :) thanks for the testing.
I will change numpy version of testing.

**Comment by jakevdp** [comment]
Thanks for the contribution, and for working out all the testing issues!
